### PR TITLE
Add framework option

### LINKED ISSUE #184: Bug: Interpolating Javascript codes creates a warn
## Info

| Tool      | Version                 |
| --------- | ----------------------- |
| Plugin    | v1.13.5                  |
| Prettier  | v2.x.x                  |
| Framework | vue |
| Node      | v14.15.3                  |
| OS        | mac           |

## Prettier config

```json
{}
```

## Input

```pug
p {{ new Date().toISOString() }}
```

## Output or Error

```bash
console.warn
    [PugPrinter:formatText]:  SyntaxError: Unexpected token 'Date'
        at o (/Users/skyatura/projects/shinigami92/plugin-pug/node_modules/prettier/parser-angular.js:46:17262)
        at Object.<anonymous>.e.parseNgInterpolation (/Users/skyatura/projects/shinigami92/plugin-pug/node_modules/prettier/parser-angular.js:46:18429)
        at s (/Users/skyatura/projects/shinigami92/plugin-pug/node_modules/prettier/parser-angular.js:46:29190)
        at Object.<anonymous>.e.parseInterpolation (/Users/skyatura/projects/shinigami92/plugin-pug/node_modules/prettier/parser-angular.js:46:29490)
        at /Users/skyatura/projects/shinigami92/plugin-pug/node_modules/prettier/parser-angular.js:46:30113
        at Object.parse (/Users/skyatura/projects/shinigami92/plugin-pug/node_modules/prettier/parser-angular.js:46:29796)
        at Object.parse (/Users/skyatura/projects/shinigami92/plugin-pug/node_modules/prettier/index.js:13625:19)
        at coreFormat (/Users/skyatura/projects/shinigami92/plugin-pug/node_modules/prettier/index.js:14899:14)
        at format (/Users/skyatura/projects/shinigami92/plugin-pug/node_modules/prettier/index.js:15131:14)
        at /Users/skyatura/projects/shinigami92/plugin-pug/node_modules/prettier/index.js:57542:12
```

### PR DESCRIPTION
**Pull request by SkyaTura**
This PR adds a new option that allows users to specify which framework is being used in the project.

This is needed to fix #184 

TODO:

- [x] Use `babel` as default parser
- [x] Add Svelte (although there is no specificity to this framework ~yet~)
- [x] Add documentation
- [x] Add framework detection
- [x] Test framework detection
- [ ] Test framework specific warnings

### CODE REVIEW

**Comment by SkyaTura** [comment]
@Shinigami92 could you please tell me if it's ok to create this option. 😄

**Comment by Shinigami92** [comment]
> @Shinigami92 could you please tell me if it's ok to create this option.

image

**Comment by Shinigami92** [comment]
I will make a more in depth review tomorrow :+1:

**Comment by Shinigami92** [comment]
I'm currently not sure if there is something special that needs to be done for `svelte` (never used it on my own)
But it could have similar syntax to Vue, I think

So what do you think about using `babel` as the default parser and only use `__ng_interpolation` if it is actually `angular`?
This would be a breaking change, but this may be okay
But the breaking change would be more _okay_ if we also directly solve #172
Do you think you could tackle this also in this PR? May only be "_if vue, then don't log_" for now

**Comment by SkyaTura** [comment]
I was thinking on avoid breaking changes for now, since it works fine (except for the warnings).

I'll ignore the vue's specific errors to solve #172

**Comment by Shinigami92** [comment]
I try to get in contact with a pug-svelte developer currently
https://discord.com/channels/457912077277855764/750401468569354431/825717140195639367

image

---

I'm just _afraid_ using the `__ng_interpolation` as the default for `'none'`, cause in theory it doesn't make sense in any way.
The `__ng_*` parsers are specially internal angular parsers, and it is a good thing if we can get rid as most as possible of them.

Don't fear a breaking change here, it is a `1.4.0` and I will later explicitly make a note of this in the changelog.

---

Beside that, I think I will create a [VuePress](https://vuepress.vuejs.org) (or similar) documentation.

**Comment by Shinigami92** [comment]
Just to get not confused ^^
You are going to solve the TODOs in this PR, will you?

I will convert this PR to a draft, just convert it back when you are finished and I can review the docs :slightly_smiling_face:

**Comment by SkyaTura** [comment]
Yes, I will as soon as I get some time to 😄

**Comment by SkyaTura** [comment]
@Shinigami92 how do I test warnings? 🤔

**Comment by Shinigami92** [comment]
> How do I test warnings? 

I thought about this a long time ago.
Short: currently there are no tests for doing that.
But IMO we should think about a test-strategy (maybe not in this issue, but in a separat one)

> Also, when you merge #191 will we maintain the README.md documentarion?

You should currently just add a documentation to the actual `main` `README.md`
I will keep track of that, when this issue is merged to `main`
The documentation website may take some more time and it's heavily in development :slightly_smiling_face: 

But yes, later we don't need to add the documentation of options in the `README.md` but the website
There is a script `yarn docs:dev` to spin up the docs on `localhost:8080`

So for now you should add a row between line `144` (`pugWrapAttributesPattern`) and `145` (`pugSingleFileComponentIndentation`), cause I think it fits best above `pugSingleFileComponentIndentation` :slightly_smiling_face:  
You may use `don't wrap lines` in your IDE to add it in there :smile:

File: src/options/pug-framework.ts [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+export const PUG_FRAMEWORK: ChoiceSupportOption = {
+	since: '1.14.0',
+	category: CATEGORY_PUG,
+	type: 'choice',
+	default: 'none',
+	description: 'Define which framework is used in the project.',
+	choices: [
+		{ value: 'none', description: 'No framework specifically.' },
+		{ value: 'angular', description: 'Uses Angular.' },
+		{ value: 'svelte', description: 'Uses Svelte' },
```
Comment (Shinigami92) [inline src/options/pug-framework.ts:14]:
```suggestion
		{ value: 'svelte', description: 'Uses Svelte.' },
```

File: src/printer.ts [hunk 2]
```diff
@@ -415,20 +433,26 @@ export class PugPrinter {
...
-								logger.warn(
-									'[PugPrinter:formatText]: Missing expected ). If you are using Vue, you can ignore this message.',
-									`code: \`${code.trim()}\``
-								);
-							} else if (error.includes('Missing expected :')) {
-								logger.warn(
-									'[PugPrinter:formatText]: Missing expected :. If you are using Vue, you can ignore this message.',
-									`code: \`${code.trim()}\``
-								);
+								if (pugFramework !== 'vue') {
```
Comment (Shinigami92) [inline src/printer.ts:436]: **thought:** Mh :thinking:  
We should really find out how `svelte` behaves here :eyes: (and the below ones)

Cause maybe we need to use `pugFramework === 'angular'` here
But I'm absolutely not sure

File: tests/issues/issue-70/issue-70.test.ts [hunk 3]
```diff
@@ -7,7 +7,12 @@ describe('Issues', () => {
 	test('should handle Framework interpolation', () => {
```
Comment (Shinigami92) [inline tests/issues/issue-70/issue-70.test.ts:7]:
```suggestion
	test('should handle angular framework interpolation', () => {
```

File: tests/issues/issue-9/issue-9.test.ts [hunk 4]
```diff
@@ -11,6 +11,8 @@ describe('Issues', () => {
 			parser: 'pug',
 			plugins: [plugin],
 
+			// @ts-expect-error
+			pugFramework: 'angular',
```
Comment (Shinigami92) [inline tests/issues/issue-9/issue-9.test.ts:15]: _cannot add comment to line 7 ^^_

`should escape quotes correctly in angular`

**Comment by Shinigami92** [comment]
@SkyaTura Just an idea, but do you think it is possible to detect which framework is in use?
e.g. read the projects package.json an look for a dependency of `angular` or `vue` or `svelte`?

And then use this as fallback if no `pugFramework` is provided?

I think we cannot write tests for that (at least not easily) so we need to just test it in a local copy/repo.
I'm thinking about using `fs` calls :thinking:

**Comment by Shinigami92** [comment]
@SkyaTura I checked how to possibly detect the projects framework.
You may read e.g. `process.env.npm_package_dependencies_vue`. The value can be `undefined` or (in my case) `'~2.6.12'`.
I think other dependencies can be checked exactly the same way :slightly_smiling_face: 

So we may use something like

```js
if (process.env.npm_package_dependencies_vue)
  pugFramework = 'vue'
else if (process.env.npm_package_dependencies_svelte)
  pugFramework = 'svelte'
else if (process.env.npm_package_dependencies_angular)
  pugFramework = 'angular'
else
  pugFramework = 'none'
```

File: src/printer.ts (hunk 2 above)
Comment (Shinigami92) [reply src/printer.ts]: We'll see if some people run into this problem
If so, they will create an issue

**Comment by Shinigami92** [comment]
@SkyaTura I'm not sure how and if we should test framework specific warnings :thinking:  
Nevertheless I worked on all other remaining tasks :smiley:  
Could you have a look and review my changes? _I know it's your branch, so you cannot give your approval. So just write a comment to let me know_

**Comment by SkyaTura** [comment]
It seems fine to me, thank you for helping. 😸

**Review by Shinigami92 — APPROVED** [review]

**Pull request merged**